sdk/core/azure-core/src/main/java/com/azure/core/util/Contexts.java

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+import com.azure.core.http.HttpPipelineCallContext;
+
+/**
+ * TODO (kasobol-msft) add docs.
+ */
+public final class Contexts {
+
+    private Contexts() {
+    }
+
+    /**
+     * TODO (kasobol-msft) add docs.
+     * @param context The context.
+     * @param progressReporter The progress reporter.
+     * @return The context.
+     */
+    public static Context setProgressReporter(Context context, ProgressReporter progressReporter) {
+        return context.addData("azure-progress-reporter", progressReporter);
+    }
```

Comment:
This API assumes that a user will only want to set one context property in a Context, which works nice in case of scenarios where the user will do `client.doFoo(...., Contexts.setProgressReporter(ctx, progressReporter)`, but I wonder what the prevalence of this approach would be, versus the need to set multiple properties in a Context, and the need at that point to extract it out into multiple lines of code. It would become something like:

```
Context ctx = Contexts.setProgressReporter(Context.NONE, progressReporter);
ctx = Contexts.setTimeout(ctx, duration);
client.doFoo(...., ctx);
```

I wonder if we should consider diving into a more fluent API from the get-go, e.g.:

```
Contexts.empty().setTimeout(...).setProgressReporter(..).create(); // from new empty Context
Contexts.with(ctx).setTimeout(...).setProgressReporter(..).create(); // from existing Context
```

We would then potentially have two classes:

```
public final class Contexts {
    private Contexts() { }

    public static Contexts empty() { return with(Context.NONE); }
    public static Contexts with(Context ctx) { return new Contexts(ctx); }

    public Contexts setTimeout(Duration timeout) { .. }
    public Contexts setProgressReporter(ProgressReporter pr) { ... }

    public Context create() { return ctx; }
}
```

The getter methods could remain on Contexts for convenience.

Reply:
Good stuff. I'll incorporate this in non-experimental PR. My goal here was to prove that progress reporting can be pulled down closer to wire.